### ISSUE TITLE
Custom Error not working using Error class extension

### ISSUE THREAD

**Issue by rcalpha** [simplesamlphp/simplesamlphp#2133]
**Specifics of your environment**
1. Are you acting as SP/IdP/proxy?: IdP
2. SimpleSAMLphp: What version are you using?: 2.2.2
3. PHP: What version are you using?: 8.2
4. Platform: unix or Windows?: linux
5. Webserver: Apache/Nginx/ISS?: Apache

**Describe the bug**

I was trying to migrate our Auth Source module which was using custom error message by my own with version 2.0.6.

I tried to use the proposed solution of issue #1867 by extending the Error class done in 2.1 and 2.2 as proposed in the pull #1870

I couldn't tried before due to limited time.

I did two class extension as proposed in our Auth Module (I'm attaching both file):

[CustomErroModuler.zip](https://github.com/user-attachments/files/15840207/CustomErroModuler.zip)

- ErrorUc extends from Error\Error
- ErrorUcCodes extends from Error\ErrorCodes

In my auth source module, which is an extension of UserPassBase class , to do a simple test, in the login method, at the beginning invoke the error with:
---------------
            throw new ErrorUc(ErrorUcCodes::BIND_SEARCH_CONNECT_ERROR);
---------------

Instead the usual
---------------
            throw new Error\Error('WRONGUSERPASS');
---------------

And I got an exception: (I attached a txt file with debug information) 
-------------------------------------------------------------
Caused by: SimpleSAML\Error\Exception: Key "BINDSEARCHCONNECTERROR" for array with keys "ACSPARAMS, ARSPARAMS, AUTHSOURCEERROR, BADREQUEST, CASERROR, CONFIG, CREATEREQUEST, DISCOPARAMS, GENERATEAUTHNRESPONSE, INVALIDCERT, LDAPERROR, LOGOUTINFOLOST, LOGOUTREQUEST, MEMCACHEDOWN, METADATA, METADATANOTFOUND, NOACCESS, NOCERT, NORELAYSTATE, NOSTATE, NOTFOUND, NOTFOUNDREASON, NOTSET, NOTVALIDCERT, PROCESSASSERTION, PROCESSAUTHNREQUEST, RESPONSESTATUSNOSUCCESS, SLOSERVICEPARAMS, SSOPARAMS, UNHANDLEDEXCEPTION, UNKNOWNCERT, USERABORTED, WRONGUSERPASS" does not exist in core:loginuserpass.twig
-------------------------------------------------------------

**To Reproduce**
I simple invoke my own Error class (extended from Error\Error): 

throw new ErrorUc(ErrorUcCodes::BIND_SEARCH_CONNECT_ERROR);

**Expected behavior**
Display the error message about the login process error. In this test it should displays: the title 'Testing Title' and description 'Testing Description'

**Screenshots or logs**
I attached the txt file with the debug info

[debug-error.txt](https://github.com/user-attachments/files/15840260/debug-error.txt)

**Additional context**

I was debugging and I found that in line 286 (version 2.2.2) of the file modules/core/src/Controller/Login.php it's invoking the own simplesamlphp ErrorCodes class instead of our own ErrorCodes extended class (ErrorUcCodes)

------------------
        $t->data['errorcodes'] = Error\ErrorCodes::getAllErrorCodeMessages();
------------------

I changed it to my own class to test:
------------------
       $t->data['errorcodes'] = Module\ucauth22\Error\ErrorUcCodes::getAllErrorCodeMessages();
------------------

And it worked! It display the custom error message

[customerror.patch](https://github.com/user-attachments/files/15840112/customerror.patch)

Maybe the solution could be:
1) A config option to set the name of custom ErrorCodes class (which I read that it will deprecate in next major version)
2) A way to create the custom ErrorCodes instance  and register from any module and invoke the error Code new getAllErrorCodeMessages method

**Comment by tvdijen** [comment]
I think instead of returning _just_ your own errors in `ErrorUcCodes.php` you have to merge yours with the ones from `\SimpleSAML\Error\ErrorCodes` ...

**Comment by tvdijen** [comment]
@monkeyiq Can you figure out the correct way to extend our error codes? And maybe write some documentation?

**Comment by rcalpha** [comment]
> I think instead of returning _just_ your own errors in `ErrorUcCodes.php` you have to merge yours with the ones from it's parent `\SimpleSAML\Error\ErrorCodes` ...
> 
> What we do have to fix on our end is that the `\SimpleSAML\Error\ErrorCodes` class should be using `static::` instead of `self::` inside it's static methods..

I could merge some custom message that maybe could be useful for many. However It doesn't make sense to have custom error option if o1ne cannot use it.

**Comment by tvdijen** [comment]
No no, I think you're misunderstanding..
My suggestion is that `ErrorUcCodes::getAllErrorCodeMessages()` is returning both your custom error and the library defaults.. So instead of `return [<your-custom-error];` I'm thinking your class should `return array_merge([<your-custom-error>], parent::<parent-method>);`

**Comment by rcalpha** [comment]
> getAllErrorCodeMessages

Ok, I undertood!!!! I didn't implemented getAllErrorCodeMessages() override this in my own class.

I'll copy that from the library default, also I will copy to override getAllErrorCodeTitles and getAllErrorCodeDescriptions as I see in the library default without change:

```
    public static function getAllErrorCodeMessages(): array
    {
        return [
            self::KEY_TITLE => self::getAllErrorCodeTitles(),
            self::KEY_DESCRIPTION => self::getAllErrorCodeDescriptions(),
        ];
    }
```

However even without this change you suggested it worked with the change I made in **modules/core/src/Controller/Login.php** (commented in the first post) which is invoking _Error\ErrorCodes::getAllErrorCodeMessages();_ instead my own error class

**Comment by tvdijen** [comment]
Why are you copying things that you do not need? You make zero sense to me, sorry..

**Comment by monkeyiq** [comment]
I'm digging into this now

**Comment by monkeyiq** [comment]
I now have a working solution to allow both custom error codes and normal error codes to work at the same time in templates. 

The differences are shown in this PR https://github.com/simplesamlphp/simplesamlphp/pull/2142 which needs to be updated to better protect the handling of the new variable added to the state which used to recreate the errorcodes class. It is likely that an explicit "yes list" will be a good idea where each custom errorcodes class registers itself and class name values that are not on that list will cause an error (if the state could in any case become contaminated).

By storing the error code class both normal errors can be raised and custom errors and the template will do the right thing in both cases. For example at lines 147 and 148 one will see an expected presentation on the login page...

https://github.com/monkeyiq/simplesamlphp/blob/b00d11cadf91e7c83024960da80412be794890bf/modules/exampleauth/src/Auth/Source/UserPass.php#L147

**Comment by rcalpha** [comment]
Hi,,,

Let's start this over again to better understand the issue.

I've made two test module:
1) testauth: Auth module to Trigger always the custom error
2) testerror: Custom ErrorCodes and Error class extended

**Comment by rcalpha** [comment]
Hi,,,

Let's start this over again to better understand the issue.

I've made two test modules to reproduce easily the issue:

* testauth: Auth module to Trigger always the custom error
* testerror: Contains two extended class from ErrorCodes and Error. ErrorCodesTest (extends from \SimpleSAML\Error\ErrorCodes) and ErrorTest (extends from \SimpleSAML\Error\ErrorCodes).
[simplesaml-test-module.tar.gz](https://github.com/user-attachments/files/15915988/simplesaml-test-module.tar.gz)

I've configured to use TestAuth for authenticate. I'm, using the 'TEST_ERROR' key for the custom test error message. I also enabled both test modules.

Without any change, it will show an exception indicating that 'TEST_ERROR" key was not found.

![screenshot orig](https://github.com/simplesamlphp/simplesamlphp/assets/3718709/a1a1ccce-4105-4d61-92dd-863c86808750)
[SimpleSAMLErrorError UNHANDLEDEXCEP.txt](https://github.com/user-attachments/files/15915985/SimpleSAMLErrorError.UNHANDLEDEXCEP.txt)

I did a trace and I found that the issue is in simplesamlphp-2.2.2/modules/core/src/Controller/Login.php which is using the original ErrorCodes class instead the custom ErrorCodes class.

I made a simple change (attaching the diff), and It displays the custom error.

![screenshot fixed](https://github.com/simplesamlphp/simplesamlphp/assets/3718709/8ff3def0-a089-4152-b15a-3ebe4cdf564c)

This is the change I made:
[simplesaml-login.patch](https://github.com/user-attachments/files/15916307/simplesaml-login.patch)

**Comment by tvdijen** [comment]
But how should the login-controller know about your custom error?  Does it make sense to use custom errors outside your custom modules?

**Comment by rcalpha** [comment]
> But how should the login-controller know about your custom error? Does it make sense to use custom errors outside your custom modules? I think your custom auth-class should catch the custom error and throw a generic one?

You are right, login-controller should not know custom error, but there should be a way that custom error message be displayed. 

It seems that monkeyiq tests could be a way.

I separated the custom error module from the auth module only for illustration purpose. It can be inside the auth module.

**Comment by monkeyiq** [comment]
The with the simplesaml-login.patch if you throw your own error code then it may appear ok. But if the module wants to throw a standard code as well then that will not work at the same time. My patch records the `get_class($e->getErrorCodes())` so that if you are using a custom errorcodes then it will be known and if not then the normal codes will be used. Your code should then be able to raise either standard or custom errors in the same method.

The other option would be for a custom errorcodes to clone the normal codes into itself and set it as the only errorcodes class in Login. Though that does not scale if two modules want to have custom error codes at the same time.
